File: January 2025/28.py
class Solution:
    def findPermutation(self, s):
        def backtrack(path, used):
            if len(path) == len(s):
                result.append("".join(path))
                return
            
            for i in range(len(s)):
                if used[i] or (i > 0 and s[i] == s[i - 1] and not used[i - 1]):
                    continue
                used[i] = True
                path.append(s[i])
                backtrack(path, used)
                path.pop()
                used[i] = False
        
        result = []
        s = sorted(s)
        used = [False] * len(s)
        backtrack([], used)
        return result

File: January 2025/28.cpp
class Solution {
public:
    vector<string> findPermutation(string &s) {
        vector<string> result;
        sort(s.begin(), s.end());
        vector<bool> used(s.size(), false);
        string path;
        backtrack(s, used, path, result);
        return result;
    }

private:
    void backtrack(string &s, vector<bool> &used, string &path, vector<string> &result) {
        if (path.size() == s.size()) {
            result.push_back(path);
            return;
        }

        for (int i = 0; i < s.size(); i++) {
            if (used[i] || (i > 0 && s[i] == s[i - 1] && !used[i - 1])) continue;
            used[i] = true;
            path.push_back(s[i]);
            backtrack(s, used, path, result);
            path.pop_back();
            used[i] = false;
        }
    }
};
